Chaining hash table for repository.Index

greatroar opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

This implements a custom chaining hash table for repository.Index.

Index merging (#2799) should be easy to add to this PR.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Alternative to #2811, with better performance on the benchmarks.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

Commit:
Chaining hash table for repository.Index

These are faster to construct but slower to access. The allocation rate
is halved, the peak memory usage almost halved compared to standard map.

Benchmark results on linux/amd64, -benchtime=3s -count=20:

name                                            old time/op    new time/op    delta
PackerManager-8                                    178ms ± 0%     178ms ± 0%     ~     (p=0.231 n=20+20)
DecodeIndex-8                                      4.54s ± 0%     4.30s ± 0%   -5.20%  (p=0.000 n=18+17)
DecodeIndexParallel-8                              4.54s ± 0%     4.30s ± 0%   -5.22%  (p=0.000 n=19+18)
IndexHasUnknown-8                                 44.4ns ± 5%    50.5ns ±11%  +13.82%  (p=0.000 n=19+17)
IndexHasKnown-8                                   48.3ns ± 0%    51.5ns ±12%   +6.68%  (p=0.001 n=16+20)
IndexAlloc-8                                       758ms ± 1%     616ms ± 1%  -18.69%  (p=0.000 n=19+19)
IndexAllocParallel-8                               234ms ± 3%     204ms ± 2%  -12.60%  (p=0.000 n=20+18)
MasterIndexLookupSingleIndex-8                     122ns ± 0%     145ns ± 9%  +18.44%  (p=0.000 n=14+20)
MasterIndexLookupMultipleIndex-8                   369ns ± 2%     429ns ± 8%  +16.27%  (p=0.000 n=20+20)
MasterIndexLookupSingleIndexUnknown-8             68.4ns ± 5%    74.9ns ±13%   +9.47%  (p=0.000 n=20+20)
MasterIndexLookupMultipleIndexUnknown-8            315ns ± 3%     369ns ±11%  +17.14%  (p=0.000 n=20+20)
MasterIndexLookupParallel/known,indices=5-8        743ns ± 1%     816ns ± 2%   +9.87%  (p=0.000 n=17+17)
MasterIndexLookupParallel/unknown,indices=5-8      238ns ± 1%     260ns ± 2%   +9.14%  (p=0.000 n=19+20)
MasterIndexLookupParallel/known,indices=10-8      1.01µs ± 3%    1.11µs ± 2%   +9.79%  (p=0.000 n=19+20)
MasterIndexLookupParallel/unknown,indices=10-8     222ns ± 0%     269ns ± 2%  +20.83%  (p=0.000 n=16+20)
MasterIndexLookupParallel/known,indices=20-8      1.06µs ± 2%    1.19µs ± 2%  +12.95%  (p=0.000 n=19+18)
MasterIndexLookupParallel/unknown,indices=20-8     413ns ± 1%     530ns ± 1%  +28.19%  (p=0.000 n=18+20)
SaveAndEncrypt-8                                  30.2ms ± 1%    30.4ms ± 0%   +0.71%  (p=0.000 n=19+19)
LoadTree-8                                         540µs ± 1%     576µs ± 1%   +6.73%  (p=0.000 n=20+20)
LoadBlob-8                                        5.64ms ± 0%    5.64ms ± 0%     ~     (p=0.883 n=18+17)
LoadAndDecrypt-8                                  5.93ms ± 0%    5.95ms ± 1%     ~     (p=0.247 n=20+19)
LoadIndex-8                                       25.1ms ± 0%    24.5ms ± 1%   -2.54%  (p=0.000 n=18+17)

name                                            old speed      new speed      delta
PackerManager-8                                  296MB/s ± 0%   296MB/s ± 0%     ~     (p=0.229 n=20+20)
SaveAndEncrypt-8                                 139MB/s ± 1%   138MB/s ± 0%   -0.71%  (p=0.000 n=19+19)
LoadBlob-8                                       177MB/s ± 0%   177MB/s ± 0%     ~     (p=0.890 n=18+17)
LoadAndDecrypt-8                                 169MB/s ± 0%   168MB/s ± 1%     ~     (p=0.227 n=20+19)

name                                            old alloc/op   new alloc/op   delta
PackerManager-8                                   91.8kB ± 0%    91.8kB ± 0%     ~     (p=0.772 n=12+19)
IndexAlloc-8                                       786MB ± 0%     400MB ± 0%  -49.04%  (p=0.000 n=20+18)
IndexAllocParallel-8                               786MB ± 0%     401MB ± 0%  -49.04%  (p=0.000 n=19+15)
SaveAndEncrypt-8                                  21.0MB ± 0%    21.0MB ± 0%   +0.00%  (p=0.000 n=19+19)

name                                            old allocs/op  new allocs/op  delta
PackerManager-8                                    1.41k ± 0%     1.41k ± 0%     ~     (all equal)
IndexAlloc-8                                        977k ± 0%      907k ± 0%   -7.18%  (p=0.000 n=20+20)
IndexAllocParallel-8                                977k ± 0%      907k ± 0%   -7.17%  (p=0.000 n=19+15)
SaveAndEncrypt-8                                    73.0 ± 0%      73.0 ± 0%     ~     (all equal)

greatroar commented:
As I said over at #2811, I'm not sure about the concurrent performance of this yet. I did see a second CPU core running the GC at 100% in the benchmarks.

greatroar commented:
Here's a quick real-world test using restic cat, as suggested by @aawsome. $id is the id of an existing tree blob in a 700GiB repo (local disk).

Current master:
```
$ /usr/bin/time ./restic cat blob $id >/dev/null 
12.77user 0.32system 0:04.39elapsed 298%CPU (0avgtext+0avgdata 705296maxresident)k
0inputs+0outputs (0major+125964minor)pagefaults 0swaps
$ /usr/bin/time ./restic -cat blob $id >/dev/null 
12.70user 0.35system 0:04.01elapsed 325%CPU (0avgtext+0avgdata 717884maxresident)k
0inputs+0outputs (0major+127844minor)pagefaults 0swaps
$ /usr/bin/time ./restic cat blob $id >/dev/null 
12.90user 0.34system 0:04.10elapsed 322%CPU (0avgtext+0avgdata 727872maxresident)k
0inputs+0outputs (0major+139900minor)pagefaults 0swaps
```

This PR:

```
$ /usr/bin/time ./restic cat blob $id >/dev/null 
13.77user 0.35system 0:04.38elapsed 322%CPU (0avgtext+0avgdata 569968maxresident)k
0inputs+0outputs (0major+136284minor)pagefaults 0swaps
$ /usr/bin/time ./restic cat blob $id >/dev/null 
13.60user 0.28system 0:03.99elapsed 348%CPU (0avgtext+0avgdata 533716maxresident)k
0inputs+0outputs (0major+123454minor)pagefaults 0swaps
$ /usr/bin/time ./restic cat blob $id >/dev/null 
13.91user 0.31system 0:04.07elapsed 349%CPU (0avgtext+0avgdata 570852maxresident)k
0inputs+0outputs (0major+137047minor)pagefaults 0swaps
```

#2811 at 64ecc669b6720fe1160731c6be895a1810347938:

```
$ /usr/bin/time ./restic -p ~/.restic/pw cat blob $id >/dev/null 
13.34user 0.22system 0:04.87elapsed 278%CPU (0avgtext+0avgdata 467384maxresident)k
0inputs+0outputs (0major+105581minor)pagefaults 0swaps
$ /usr/bin/time ./restic -p ~/.restic/pw cat blob $id >/dev/null 
13.22user 0.27system 0:04.18elapsed 322%CPU (0avgtext+0avgdata 478512maxresident)k
0inputs+0outputs (0major+109278minor)pagefaults 0swaps
$ /usr/bin/time ./restic -p ~/.restic/pw cat blob $id >/dev/null 
13.27user 0.24system 0:04.15elapsed 325%CPU (0avgtext+0avgdata 468200maxresident)k
0inputs+0outputs (0major+104110minor)pagefaults 0swaps
```

Here, this PR so far achieves a 23% reduction in average memory use, while #2811 gets a 35% reduction. There's a few constants left to tune, I'll see if can lower.

greatroar commented:
Nope, tuning constants doesn't make much of a difference. I tried again on a smaller, 210GiB repo, and there it's 32% for #2811, 18% for this PR.

greatroar commented:
Following @aawsome's advice, I reorganised the duplicates and increased the maximum load factor to make more efficient use of memory. In the first restic cat tests, I now get a 32% reduction and I support I can get up to the level of #2811 by increasing it further.

greatroar commented:
Test coverage for indexmap.go is 98.1%, or one return statement not covered.

aawsome commented on internal/repository/index.go:
```diff
@@ -214,35 +187,30 @@ func (idx *Index) Lookup(id restic.ID, tpe restic.BlobType) (blobs []restic.Pack
+			list = append(list, idx.toPackedBlob(e, restic.DataBlob))
 		}
-	}
+		return true
```
This makes `foreach` return after the first blob in this pack. It should loop over all entries and append all blobs with matching pack ID.

greatroar replied:
No, the callback returns true to signal that it wants to continue. When this is set to false, TestIndexUnserialize fails.

aawsome replied:
Ah - I missed the `!` in `foreach`. Then it's ok!

aawsome commented on internal/repository/index.go:
```diff
@@ -214,35 +187,30 @@ func (idx *Index) Lookup(id restic.ID, tpe restic.BlobType) (blobs []restic.Pack
-	}
+		return true
+	})
+	idx.tree.foreach(func(e *indexEntry) bool {
```
can be replaced by a loop when using `[]*indexMap`

aawsome commented on internal/repository/index.go:
```diff
@@ -302,15 +267,21 @@ func (idx *Index) Each(ctx context.Context) <-chan restic.PackedBlob {
+
+		typ = restic.DataBlob
+		idx.data.foreach(send)
+		typ = restic.TreeBlob
```
see above about `[]*indexMap`

aawsome commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+
+const (
+	growthFactor = 2 // Must be a power of 2.
+	maxLoad      = 4 // Max. number of entries per bucket.
```
With `maxLoad = 4` we have minimum 2 and maximum 4 entries per bucket. For each bucket we have to store a pointer of 8 bytes. Hence this is an overhead of 2 to 4 bytes per Entry. I would advise to choose a slightly higher number, e.g. 6 or 8. This will decrease Lookup speed a bit but may easily save another byte per entry. IMO a much higher Lookup time can be afforded once we merge indexes into one (actually Lookup loops over maybe hundreds of index files).

This proposed change however can easily be postponed to the following merge PR :wink:

greatroar replied:
Then I say we postpone. We still need good lookup performance for #2787 in the meantime.

aawsome reviewed (changes requested):
Looks good so far; I have just a few small changes to propose.

Side note: How do I get information about test code coverage?

aawsome commented:
I would also add a changelog, as this is another massive improvement with respect to memory usage!

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+
+func (m *indexMap) hash(id restic.ID) uint {
+	h := uint(binary.LittleEndian.Uint64(id[:8]))
+	return h & uint(len(m.buckets)-1)
```
In a previous version, I used %, which was killing performance :)

Re: computing all over, subtracting 1is a single instruction, which can also run in parallel with the load that LittleEndian.Uint64 compiles to.

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+// foreach calls fn for all entries in the map, until fn returns false.
+func (m *indexMap) foreach(fn func(*indexEntry) bool) {
```
Because one of the callers wants to exit when its context expires.

greatroar commented on internal/repository/index.go:
```diff
@@ -140,7 +114,7 @@ var IndexFull = func(idx *Index) bool {
 	debug.Log("checking whether index %p is full", idx)
 
-	blobs := len(idx.blob)
+	blobs := idx.data.len()
```
Good catch, fixed!

greatroar commented:
I intend to merge the changelog entry with the one for #2781, because to a user, they're the same thing. Is that ok with you?

aawsome commented:
> I intend to merge the changelog entry with the one for #2781, because to a user, they're the same thing. Is that ok with you?

Sure - I think it is good idea to tell the overall improvement to users.

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+
+// get returns the first entry for the given id.
+func (m *indexMap) get(id restic.ID) *indexEntry {
+	if len(m.buckets) == 0 {
```
I'm relying on the branch predictor to figure out that this branch is seldom taken, to save some memory in the case where it actually is. I doubt that there will be much of a speedup, if any, based on experience with lazy initialization in other projects (and even the stdlib does it quite often).

aawsome replied:
All right! Also didn't expect much of a speedup but had in mind that lookups are really called *very* often.
But with your explanation I think we should keep it as it is.

greatroar replied:
I benchmarked a version without lazy init and it turns out that MasterIndexLookupSingleIndexUnknown actually runs 10% slower without lazy init. I'm not sure why that is. The rest of the benchmarks run at the same speed.

greatroar commented:
Re: test coverage,

```
go test -coverprofile=cov
go tool cover -html cov
```

greatroar commented on internal/repository/index.go:
```diff
@@ -19,40 +19,33 @@ import (
-	treePacks  restic.IDs
+	m         sync.Mutex
+	data      indexMap
+	tree      indexMap
```
I hear you, but I also like the fact that an invalid blob type gives a nil map from mapFor now. That leads to an immediate crash without validation code.

aawsome replied:
same in #2811 with `[]*indexMap` ;-)
There `InvalidBlob` gets an index of 0 (which is set to `nil` by `make([]*indexMap, len(allBlobTypes)+1)` and indices 1 (`DataBlob`) and 2 (`TreeBlob`) are initialized to a indexMap. All other BlobTypes would throw an "out of index error" unless added to `allBlobTypes`

aawsome replied:
Actually we should consider moving `AllBlobTypes` variable to `internal/restic/blob.go`...

greatroar replied:
I tried doing this with InvalidBlob removed (it's never used anyway) and a sentinel added, so

```go
const (
	DataBlob BlobType = iota
	TreeBlob
	BlobTypeMax
)
```

and a field `byType [restic.BlobTypeMax]*indexMap`, but then the tests fail.

I'll leave this unresolved for now.

greatroar replied:
Done. I added a sentinel NumBlobTypes to restic.BlobType, leaving InvalidBlob alone (I saw that one of @aawsome's PRs uses it).

aawsome commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+	old := m.buckets
+	m.buckets = make([]*indexEntry, growthFactor*len(m.buckets))
+
+	for _, e := range old {
```
Maybe we should consider to run this loop concurrently? Within one iteration, only buckets for new hashes are accessed and these are completely isolated when using a growFactor of two. I'm just fearing that growing large indexes will keep only 1 CPU occupied for quite some time...

greatroar replied:
You mean like so?

```go
var sem = make(chan struct{}, runtime.NumCPU())

func (m *indexMap) grow() {
	old := m.buckets
	m.buckets = make([]*indexEntry, growthFactor*len(m.buckets))

	if len(m.buckets) < 8192 {
		for _, e := range old {
			for e != nil {
				h := m.hash(e.id)
				next := e.next
				e.next = m.buckets[h]
				m.buckets[h] = e
				e = next
			}
		}
		return
	}

	for _, e := range old {
		m.wg.Add(1)
		sem <- struct{}{}

		go func(e *indexEntry) {
			for e != nil {
				h := m.hash(e.id)
				next := e.next
				e.next = m.buckets[h]
				m.buckets[h] = e
				e = next
			}

			<-sem
			m.wg.Done()
		}(e)
	}

	m.wg.Wait()
	m.wg = sync.WaitGroup{}
}
```

This does:

```
name          old time/op    new time/op    delta
DecodeIndex-8    16.1µs ± 4%    16.0µs ± 3%     ~     (p=0.571 n=12+10)
IndexAlloc-8      629ms ± 2%     827ms ± 1%  +31.55%  (p=0.000 n=11+10)
```

I can see it using at most 3.5 of my 8 cores in top. I suppose it's too memory-bound.

We could try a preallocated goroutine pool, but this function is mostly called from worker pools doing SaveBlob or DecodeIndex anyway.

greatroar replied:
A nil check before ch <- e helps a bit, but not much.

```
name           old time/op    new time/op    delta
DecodeIndex-8    16.1µs ± 4%    15.8µs ± 3%     ~     (p=0.143 n=12+9)
IndexAlloc-8      629ms ± 2%     803ms ± 2%  +27.76%  (p=0.000 n=11+10)
```

aawsome replied:
Too much of communication :flushed: 

Using
```
cpus := runtime.NumCPU()
	var wg sync.WaitGroup
	wg.Add(cpus)
	for cpu := 0; cpu < cpus; cpu++ {
		go func(cpu int) {
			for i := cpu; i < len(m.buckets); i += cpus {
				b := m.buckets[i]
				for b != nil {
					index := b.key.Hash() & m.mask
					next := b.next
					b.next = newBuckets[index]
					newBuckets[index] = b
					b = next
				}
			}
			wg.Done()
		}(cpu)
	}
	wg.Wait()
```
in my code (but should be equal to your implementation) gives me:
```
alex@thinkpad:~/go/src/github.com/restic/restic/internal/repository$ benchstat old new
name          old time/op    new time/op    delta
IndexAlloc-4     1.06s ± 9%     0.90s ± 3%  -15.34%  (p=0.000 n=10+10)

name          old alloc/op   new alloc/op   delta
IndexAlloc-4     410MB ± 0%     410MB ± 0%     ~     (p=0.315 n=10+10)

name          old allocs/op  new allocs/op  delta
IndexAlloc-4     2.63M ± 0%     2.63M ± 0%     ~     (p=0.156 n=10+9)
```

greatroar replied:
Nice! What size cap did you use to decide on sequential vs. parallel?

aawsome replied:
In fact I didn't use any. My judgement is that optimizing growing for small tables is not worth to optimize for, but I didn't benchmark it, though.

greatroar replied:
7897775f64ff3a31e36adc301a94938791294c3e.

greatroar commented on internal/repository/index.go:
```diff
@@ -19,40 +19,33 @@ import (
 //
 // To save N index entries, we therefore need:
-// N * OF * (33 + 16) bytes + N * 32 bytes / BP = N * 78 bytes
+// N * (56 + 2) bytes + N * 32 bytes / BP = N * 62 bytes.
```
@aawsome, does this strike you as correct? I feel I might have made a mistake in estimating the average load factor.

aawsome replied:
I think that the calculations of the "old" implementation cannot be taken 1:1 to see the memory usage of this "new" implementation.

For the old implementation I argued that index entries of 49 bytes needed to be stored in a golang standard hashmap. This is usually partly filled and I assumed 2/3 filling (or overhead factor OF=1.5). What is completely neglected is the extra overhead that comes from the map, i.e. the pointers to each bucket and maybe even more overhead when the map is growing. So the `N * OF * 49 bytes` are the contents within the map but without the organizational overhead the map uses itself.

For this "new" implementaion, speaking in the same context would be:
We store 48 bytes per entry and there is no overhead. :smile: This would lead to `N * 52 bytes` total usage.

However, we also do have some "organizational" overhead here: 1 pointer (8 bytes) per entry plus 1 pointer per bucket. The number of buckets is between `N/2` and `N/4`, so in total the overhead is smaller than `N * 12 bytes` (+ some overhead not depending on the number of entries, but this we can neglect).

So for this new implementation the total memory is less than `64 bytes` per entry *including any overhead*.

MichaelEischer reviewed (commented):
I like the hash table design in general. I've been mostly pondering over two points:

How large is the GC overhead? The indexEntries and buckets arrays add a massive amount of pointers which the garbage collector has to follow during GC. It might be comparable to the previous index implementation which stored slices in the index hashmap (which contain a pointer). So, it might turn out to be no problem at all. 

What are the main design choices and are there alternatives? From what I see the main points are to store pointers in the bucket map, allocate blocks of index entries (but still use them separately), chain entries for conflict resolution and use rather long chains with up to four elements on average. The first two points allow for a rather smooth increase in memory usage when a hashmap grows. The latter two points keep the overhead of bucket map pointers relatively low. (And to keep multiple entries per key instead of overwriting the old value).

Storing the index entries in the bucket array would cause the large memory usage spikes while growing the hashmap so that's not an option (unless we want to split the bucket array into parts and adjust it part by part). Just allocating individual entries leads to a lot of internal memory fragmentation and allocating everything at once doesn't work either. Using shorter chains probably isn't that helpful and an average chain length of 4 is much smaller than the usual number of indexes that currently exist in the master index. This leaves open addressing/probing as alternative to chaining. It would remove the next pointer from the indexEntry, however, in exchange the bucket array would have to be larger than the current number of entries. This becomes a problem while growing the bucket array as even with an optimal load factor and a very small grow factor this would require keeping more than two pointers per entry (three for a growth factor of 2), which is a lot more than the chaining variant needs. The main benefit of probing would be that the hashmap wouldn't have to chase linked list entries located at random places in memory but could get away with an access pattern that offers better locality.

I was wondering whether it would be possible to move/reorder entries in the grow method in such a way that the linked elements of a bucket are laid out consecutively in memory. That would at keep the last half of the chain elements next to each other. But half of the chain elements are just two index entries, so it's probably not worth the trouble...

aawsome commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func (m *indexMap) hash(id restic.ID) uint {
+	h := uint(binary.LittleEndian.Uint64(id[:8]))
```
Note that id already is a SHA256 hash. It should pretty hard to create lot's of contents efficiently with identical first bytes. If it was easy, bitcoin would be worthless immediately.

Of course you could "manipulate" your date (e.g. index files) in some sense such that you do not save SHA256 hashes. However then you have restic processing garbage (and it will abort sooner or later) and hopefully don't care about performance :wink:

greatroar replied:
Thinking out loud again.

All hashes are susceptible to birthday attacks: you can find a collision for the first *k* bits in 2^(*k*/2) time. To turn that into a hash flooding attack, you'd need to generate *n* × 2^(*k*/2) colliding strings for a chain of length *n*, each of length lg(*n*) - 1 bits, on average. To create a single linked list (*n* = 2^*k*), that comes down to (*k* - 1) 2^(1.5 *k*) or 23 × 2^36 work for 24 bits of hash. That's 1.47 TiB of data, disregarding the fact that only 1/8 bit strings are also possible byte strings.

aawsome replied:
> Thinking out loud again.
> 
> All hashes are susceptible to birthday attacks: you can find a collision for the first _k_ bits in 2^(_k_/2) time. To turn that into a hash flooding attack, you'd need to generate _n_ × 2^(_k_/2) colliding strings for a chain of length _n_, each of length lg(_n_) - 1 bits, on average. To create a single linked list (_n_ = 2^_k_), that comes down to (_k_ - 1) 2^(1.5 _k_) or 23 × 2^36 work for 24 bits of hash. That's 1.47 TiB of data, disregarding the fact that only 1/8 bit strings are also possible byte strings.

I couldn't follow your computations without thinking in more depth about it. 
However I have the feeling that your argumentation may be wrong: When you argue with birthday attacks, you calculate the work to get an arbitrary hash collusion, i.e. two pieces of data with identical hashes. When looking for the third (or more) data piece, you are no longer free about which hash this should have, but you must find some data with that given hash. So finding lots of data with arbitrary identical hash is almost as hard as finding lots of data with a given hash.

greatroar replied:
Then again, even for very small *k* it's possible to create a significant slowdown, esp. if you can generate multiple chains, which may be easier. I'll put in a randomized xxhash.

greatroar replied:
Following the advice [here](https://crypto.stackexchange.com/q/6408), I've switched to SipHash instead. Maybe xxhash with SHA-256 inputs is safe, but I don't want to be the roll-your-own-crypto guy. My benchmarks show that siphash.Hash is only about 1.5× slower than xxhash.Sum64. This does have the effect of slowing some operations down compared to current master, but not by a great deal.

    PackerManager-8                             181ms ± 2%     179ms ± 1%   -1.25%  (p=0.005 n=10+10)
    DecodeIndex-8                              13.6µs ± 2%    18.1µs ± 3%  +32.46%  (p=0.000 n=10+10)
    IndexHasUnknown-8                          43.4ns ± 5%    52.3ns ±17%  +20.43%  (p=0.000 n=10+10)
    IndexHasKnown-8                            49.2ns ± 3%    50.1ns ± 8%     ~     (p=0.088 n=10+9)
    IndexAlloc-8                                809ms ± 2%     576ms ± 2%  -28.79%  (p=0.000 n=9+10)
    MasterIndexLookupSingleIndex-8              122ns ± 0%     152ns ± 4%  +25.00%  (p=0.000 n=6+10)
    MasterIndexLookupMultipleIndex-8            379ns ± 2%     445ns ±10%  +17.38%  (p=0.000 n=10+10)
    MasterIndexLookupSingleIndexUnknown-8      68.9ns ± 4%    70.9ns ±11%     ~     (p=0.304 n=9+10)
    MasterIndexLookupMultipleIndexUnknown-8     317ns ± 1%     347ns ± 4%   +9.28%  (p=0.000 n=9+9)
    SaveAndEncrypt-8                            101ns ± 2%      95ns ± 0%   -5.35%  (p=0.000 n=10+8)
    LoadTree-8                                  350µs ±12%     452µs ± 2%  +29.36%  (p=0.000 n=10+10)
    LoadBlob-8                                 5.78ms ± 2%    5.66ms ± 0%   -2.22%  (p=0.000 n=10+9)
    LoadIndex-8                                32.8ms ± 2%    31.7ms ± 2%   -3.43%  (p=0.000 n=10+10)
    
    name                                     old speed      new speed      delta
    PackerManager-8                           291MB/s ± 2%   295MB/s ± 1%   +1.26%  (p=0.005 n=10+10)
    SaveAndEncrypt-8                         41.7TB/s ± 2%  44.0TB/s ± 1%   +5.49%  (p=0.000 n=10+9)
    LoadBlob-8                                173MB/s ± 2%   177MB/s ± 0%   +2.25%  (p=0.000 n=10+9)
    
    name                                     old alloc/op   new alloc/op   delta
    PackerManager-8                            91.5kB ± 0%    91.5kB ± 0%     ~     (p=0.839 n=10+10)
    IndexAlloc-8                                786MB ± 0%     400MB ± 0%  -49.06%  (p=0.000 n=10+10)
    SaveAndEncrypt-8                            1.00B ± 0%     1.00B ± 0%     ~     (all equal)
    
    name                                     old allocs/op  new allocs/op  delta
    PackerManager-8                             1.41k ± 0%     1.41k ± 0%   -0.06%  (p=0.040 n=10+10)
    IndexAlloc-8                                 977k ± 0%     1116k ± 0%  +14.28%  (p=0.000 n=10+10)
    SaveAndEncrypt-8                             0.00           0.00          ~     (all equal)

greatroar replied:
> When you argue with birthday attacks, you calculate the work to get an arbitrary hash collusion, i.e. two pieces of data with identical hashes. When looking for the third (or more) data piece, you are no longer free about which hash this should have, but you must find some data with that given hash. So finding lots of data with arbitrary identical hash is almost as hard as finding lots of data with an a priori given hash.

Sorry, I hadn't seen this message when I posted the previous comments.

Regardless of my previous reasoning, which may well be wrong, you don't even need to create such a long chain. You can trivially generate multiple chains of a few hundred blobs by looking for collisions in the upper bits of SHA-256. If you pad those with a big enough blob to meet minPackSize, you can generate an index that is two orders of magnitude slower than anything else.

MichaelEischer replied:
Go 1.14 has a new package that looks like a perfect match: https://golang.org/pkg/hash/maphash/ . However, it's not available on older Go versions.

greatroar replied:
Despite maphash being 10-17% faster than siphash on its own, operations using it become slower. I removed the parallel grow to get rid of copying and ensured that I did not introduce any allocations in the hot paths, but I get these benchmark results:

```
name                                     old time/op    new time/op    delta
PackerManager-8                             179ms ± 1%     179ms ± 1%     ~     (p=0.190 n=10+10)
DecodeIndex-8                              18.1µs ± 3%    16.0µs ± 4%  -11.70%  (p=0.000 n=10+10)
IndexHasUnknown-8                          52.3ns ±17%    55.5ns ±13%     ~     (p=0.324 n=10+10)
IndexHasKnown-8                            50.1ns ± 8%    54.5ns ±11%   +8.78%  (p=0.000 n=9+10)
IndexAlloc-8                                576ms ± 2%     756ms ± 4%  +31.18%  (p=0.000 n=10+10)
MasterIndexLookupSingleIndex-8              152ns ± 4%     158ns ± 7%     ~     (p=0.075 n=10+10)
MasterIndexLookupMultipleIndex-8            445ns ±10%     469ns ±10%     ~     (p=0.105 n=10+10)
MasterIndexLookupSingleIndexUnknown-8      70.9ns ±11%    76.5ns ± 7%   +7.87%  (p=0.007 n=10+9)
MasterIndexLookupMultipleIndexUnknown-8     347ns ± 4%     378ns ± 7%   +9.10%  (p=0.000 n=9+10)
SaveAndEncrypt-8                           95.2ns ± 0%    99.9ns ± 0%   +4.91%  (p=0.000 n=8+7)
LoadTree-8                                  452µs ± 2%     491µs ± 3%   +8.63%  (p=0.000 n=10+10)
LoadBlob-8                                 5.66ms ± 0%    5.66ms ± 0%   +0.10%  (p=0.036 n=9+8)
LoadIndex-8                                31.7ms ± 2%    32.0ms ± 3%     ~     (p=0.243 n=10+9)
```

aawsome commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,167 @@
+		return
+	}
+
+	m.wg.Add(cpus)
```
The total speedup of the index creation is around 20%, but this anyway should be dominated by inserting single entries into the hashmap and not the very singular growing steps. I think the plain growing step will use  full CPUs. 

Do you think the added complexity of slow migration is worth the benefit?

greatroar replied:
With SipHash in place, the speedup is a tiny bit more:

```
name          old time/op    new time/op    delta
IndexAlloc-8     767ms ± 0%     576ms ± 2%  -24.84%  (p=0.000 n=8+10)
```

That's compared to a version with SipHash, but without parallel grow.

greatroar replied:
Incremental growing would require metadata that takes up memory. Every add call needs to find two entries in the old table, which can now only be done by a linear sweep.

aawsome commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+	// Allocating in batches means that we get closer to optimal space usage,
+	// as Go's malloc will overallocate for structures of size 56 (indexEntry
+	// on amd64).
+	const entryAllocBatch = 8
```
I'm also curious about the effect to performance when changing this.

greatroar replied:
Here's an interesting result. This is with SipHash.

```go
name                                     old time/op    new time/op    delta
PackerManager-8                             183ms ± 1%     178ms ± 0%    -2.35%  (p=0.000 n=12+5)
DecodeIndex-8                              16.1µs ± 4%    37.6µs ± 3%  +133.43%  (p=0.000 n=12+5)
IndexHasUnknown-8                          26.1ns ± 6%    54.6ns ±17%  +109.44%  (p=0.001 n=10+5)
IndexHasKnown-8                            26.3ns ±29%    52.0ns ± 4%   +97.67%  (p=0.000 n=11+5)
IndexAlloc-8                                629ms ± 2%     534ms ± 2%   -15.09%  (p=0.000 n=11+5)
MasterIndexLookupSingleIndex-8              131ns ± 7%     147ns ± 2%   +11.86%  (p=0.000 n=12+5)
MasterIndexLookupMultipleIndex-8            302ns ±11%     436ns ± 7%   +44.49%  (p=0.000 n=12+5)
MasterIndexLookupSingleIndexUnknown-8      51.7ns ±30%    70.3ns ± 4%   +35.97%  (p=0.000 n=12+5)
MasterIndexLookupMultipleIndexUnknown-8     206ns ± 6%     354ns ± 6%   +71.57%  (p=0.000 n=11+5)
SaveAndEncrypt-8                           70.2ns ± 1%    97.1ns ± 1%   +38.33%  (p=0.000 n=12+5)
LoadTree-8                                  456µs ± 2%     452µs ± 1%      ~     (p=0.104 n=12+5)
LoadBlob-8                                 5.75ms ± 1%    5.70ms ± 1%    -0.90%  (p=0.006 n=12+5)
LoadAndDecrypt-8                           7.10ms ± 3%    6.98ms ± 2%    -1.68%  (p=0.037 n=12+5)
LoadIndex-8                                31.2ms ± 4%    31.6ms ± 3%      ~     (p=0.328 n=12+5)
```

However, I also see in top that the test process runs at around 235% CPU during lookup (!) benchmarks. I reran the tests with entryAllocBatch = 8 and it peaks at over 300%, staying comfortably above 260%. I suspect this is the concurrent GC at work. Allocating in bigger batches seems to trade CPU time spent allocating for CPU time spent in GC.

I think we need parallel benchmarks. I'll see if I can write a few, if nobody beats me to it.

MichaelEischer replied:
Which batch size is `old` and `new`? The Lookup tests probably spend most of their time in `createRandomIndex()` (0.6 seconds per benchmark call / 1 second with `-cpu 1` flag for a alloc batch size of 8). The test driver aims at an 1 second long benchmark execution. So most time is spent creating the index, which is probably what is visible in top.

greatroar commented:
> How large is the GC overhead? The indexEntries and buckets arrays add a massive amount of pointers which the garbage collector has to follow during GC. It might be comparable to the previous index implementation which stored slices in the index hashmap (which contain a pointer). So, it might turn out to be no problem at all.

I've yet to benchmark this.

> What are the main design choices and are there alternatives? From what I see the main points are to store pointers in the bucket map, allocate blocks of index entries (but still use them separately), chain entries for conflict resolution and use rather long chains with up to four elements on average. The first two points allow for a rather smooth increase in memory usage when a hashmap grows. The latter two points keep the overhead of bucket map pointers relatively low. (And to keep multiple entries per key instead of overwriting the old value).

The main alternatives that I considered are a probing table, which has the problems you spotted, or cuckoo hashing, which can handle very high load factors. The latter looked to be much more complicated and I quickly gave up on it.

Thinking out load, there is perhaps a middle ground: a probing table that stores summaries of the keys and a pointer to the entry, viz.

```go
type bucket struct {
	hash  uint
	entry *indexEntry // no longer needs a next pointer
}
```

Say we set the maximum load factor at 80% and grow in powers of two. Then the table will be 60% full on average and the worst case overhead will be 16*(1+2/3)*n = 26.7n bytes. Dropping the next pointer saves eight bytes, and we pay 74.7 bytes per entry.

The benefit would be the access pattern. The buckets get packed four to a cache line (on amd64/arm64), so lookup might be faster despite the probing. Insert would no longer be worst-case O(1), but nearly all the pointer indirection is gone.

I'll see if I can get this working.

> I was wondering whether it would be possible to move/reorder entries in the grow method in such a way that the linked elements of a bucket are laid out consecutively in memory. That would at keep the last half of the chain elements next to each other. But half of the chain elements are just two index entries, so it's probably not worth the trouble...

Also, the second entry in a pair necessary spills over into the next cache line because the SHA-256 is 32 bytes.

aawsome commented:
> ```go
> type bucket struct {
> 	hash  uint
> 	entry *indexEntry // no longer needs a next pointer
> }
> ```
> 
> Say we set the maximum load factor at 80% and grow in powers of two. Then the table will be 60% full on average and the worst case overhead will be 16*(1+2/3)*n = 26.7n bytes. Dropping the next pointer saves eight bytes, and we pay 74.7 bytes per entry.

No need to save the hash as it can be easily computed.
If you save `*indexEntry` in your bucket and use probing, you gain 8 bytes per index entry. In your actual implementation your buckets need between 2 and 4 bytes per index entry. So if you could use a probing table with load factor between 2/3 and 4/5 (that should be the maximum load factor!), you will get the identical memory usage. However, if you grow a hash table with load factor 80% by doubling its buckets, it will result in a hash table with load factor 40%. Then the buckets will use 5/2 * 8 bytes = 20 bytes per entry.

And that is the main drawback of hashmaps with load factor smaller than 1 - you get memory spikes when growing...

greatroar commented:
> No need to save the hash as it can be easily computed.

The hash is saved so four of them will be in the L1 cache before we follow a single pointer. That's also the only benefit of this proposed scheme.

greatroar commented on changelog/unreleased/pull-2781:
```diff
@@ -1,6 +1,8 @@
 
 We've improved how the index is stored in memory.
-This change reduces memory usage for large repositories by about 30-40%.
+This change can reduces memory usage for large repositories by up to 50%
```
Done.

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+	h := m.hash(id)
+	e := m.newEntry()
+	e.id = id
+	e.next = m.buckets[h]
```
Done.

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,140 @@
+// It allows storing multiple entries with the same key.
+//
+// A chained table has less memory overhead than a probing one (like the
+// standard map), because only a pointer array needs to be resized when the
```
Changed both.

MichaelEischer commented:
> > No need to save the hash as it can be easily computed.
> 
> The hash is saved so four of them will be in the L1 cache before we follow a single pointer. That's also the only benefit of this proposed scheme.

The standard go hash maps have a tophash field in their buckets which seems to be the standard way to avoid following unnecessary pointers/comparing entries: The field just contains the top 8 bits of an entry's hash (ignoring a few reserved values). That way after checking just that single byte you only have a 1/256 chance for a hash collision.

I think we can live with a slightly slower hashmap in exchange for the better memory efficiency. After all a merged hashmap is fast enough that the lookup-heavy `check` command only spends a tiny fraction of it's time there.

aawsome commented:
> I think we can live with a slightly slower hashmap in exchange for the better memory efficiency. After all a merged hashmap is fast enough that the lookup-heavy `check` command only spends a tiny fraction of it's time there.

Completely agree! The most important point is that we use a hashmap that does not have significant overhead even in edge cases like directly after growing which is not too slow.

This proposed solution has the following properties:

- it uses much less memory and allocations compared to the standard go hashmap
- it handles duplicates very smoothly
- it has a very "constant" total overhead (between 10 and 12 byte per entry)
- it is perfect to implement merging as this can be done almost without newly allocating memory
- it is already faster than than the actual implementation in some tests and way faster once the indexes are merged 

So if we should spend more energy in working on this solution IMO the only point worth is to benchmark and test if we missed some severe performance issues in nasty cases (like the GC topic or with concurrency)

Commit:
Parallel index benchmarks + benchmark optimizations

createRandomIndex was using the global RNG, which locks on every call
It was also using twice as many random numbers as necessary and doing
a float division in every iteration of the inner loop.

BenchmarkDecodeIndex was using too short an input, especially for a
parallel version. (It may now be using one that is a bit large.)

Results on linux/amd64, -benchtime=3s -count=20:

name                                     old time/op    new time/op       delta
PackerManager-8                             178ms ± 0%        178ms ± 0%           ~     (p=0.165 n=20+20)
DecodeIndex-8                              13.6µs ± 2%  4539886.8µs ± 0%  +33293901.38%  (p=0.000 n=20+18)
IndexHasUnknown-8                          44.4ns ± 7%       44.4ns ± 5%           ~     (p=0.873 n=20+19)
IndexHasKnown-8                            49.2ns ± 3%       48.3ns ± 0%         -1.86%  (p=0.000 n=20+16)
IndexAlloc-8                                802ms ± 1%        758ms ± 1%         -5.51%  (p=0.000 n=20+19)
MasterIndexLookupSingleIndex-8              124ns ± 1%        122ns ± 0%         -1.41%  (p=0.000 n=20+14)
MasterIndexLookupMultipleIndex-8            373ns ± 2%        369ns ± 2%         -1.13%  (p=0.001 n=20+20)
MasterIndexLookupSingleIndexUnknown-8      67.8ns ± 3%       68.4ns ± 5%           ~     (p=0.753 n=20+20)
MasterIndexLookupMultipleIndexUnknown-8     316ns ± 3%        315ns ± 3%           ~     (p=0.846 n=20+20)
SaveAndEncrypt-8                           30.5ms ± 1%       30.2ms ± 1%         -1.09%  (p=0.000 n=19+19)
LoadTree-8                                  527µs ± 1%        540µs ± 1%         +2.37%  (p=0.000 n=19+20)
LoadBlob-8                                 5.65ms ± 0%       5.64ms ± 0%         -0.21%  (p=0.000 n=19+18)
LoadAndDecrypt-8                           7.07ms ± 2%       5.93ms ± 0%        -16.15%  (p=0.000 n=19+20)
LoadIndex-8                                32.1ms ± 2%       25.1ms ± 0%        -21.64%  (p=0.000 n=20+18)

name                                     old speed      new speed         delta
PackerManager-8                           296MB/s ± 0%      296MB/s ± 0%           ~     (p=0.159 n=20+20)
SaveAndEncrypt-8                          138MB/s ± 1%      139MB/s ± 1%         +1.10%  (p=0.000 n=19+19)
LoadBlob-8                                177MB/s ± 0%      177MB/s ± 0%         +0.21%  (p=0.000 n=19+18)
LoadAndDecrypt-8                          141MB/s ± 2%      169MB/s ± 0%        +19.24%  (p=0.000 n=19+20)

name                                     old alloc/op   new alloc/op      delta
PackerManager-8                            91.8kB ± 0%       91.8kB ± 0%           ~     (p=0.826 n=19+12)
IndexAlloc-8                                786MB ± 0%        786MB ± 0%         +0.01%  (p=0.000 n=20+20)
SaveAndEncrypt-8                           21.0MB ± 0%       21.0MB ± 0%         -0.00%  (p=0.012 n=20+19)

name                                     old allocs/op  new allocs/op     delta
PackerManager-8                             1.41k ± 0%        1.41k ± 0%           ~     (all equal)
IndexAlloc-8                                 977k ± 0%         977k ± 0%         +0.01%  (p=0.022 n=20+20)
SaveAndEncrypt-8                             73.0 ± 0%         73.0 ± 0%           ~     (all equal)

greatroar commented:
I just did the following with master and this PR:

```
restic init
echo 3 | sudo tee /proc/sys/vm/drop_caches # flush kernel page cache
/usr/bin/time restic backup 83GiBofrandomdata/
```

This gives 221MiB max resident vs. 232 on master, while the elapsed time is practically the same. That's a small margin, but note that just running `/usr/bin/time restic` with no arguments takes 10MiB and the indices that this produces are only 8MiB on disk. So at the very least there is no regression.

MichaelEischer commented:
> > How large is the GC overhead? The indexEntries and buckets arrays add a massive amount of pointers which the garbage collector has to follow during GC. It might be comparable to the previous index implementation which stored slices in the index hashmap (which contain a pointer). So, it might turn out to be no problem at all.
> 
> I've yet to benchmark this.

I've just run a rather simple test with the check command (with #2328 merged), which issues lots of `LookupSize` calls. The old index (the hashmap with slices) seems to be the fastest regarding index lookup performance: 12.5 seconds + 3s background GC marking. The current master takes 16.5s without noticeable GC marking time. And the chaining hash map takes 24s + 4s background GC marking. Interestingly most of the time is spent in `memeqbody` (13s), that is `e.id == id`. The hash function comes in next with 3s for siphash. And the for-loop in `get` also takes around 3s according to pprof.

With #2818 the lookup costs inside`LookupSize` (should) drop a lot, so this slightly more expensive index lookup should be no problem.

`/usr/bin/time/` shows me the following maximum resident set size for my check runs:
644MB for this PR, 715MB for the current index on master and 891MB for the old index.

greatroar commented:
My money is on the tophash field for those 13s. Map calls memequal too, but we do four calls on average for a lookup, map one plus some byte comparisons. My attempts to optimize the comparison have not born much result.

MichaelEischer commented:
@greatroar The standard hashmap uses an `[8]uint8` array to store the tophash for the entries in a bucket. That way all tophash fields are loaded at once into the cache and thus it's really cheap to check whether any of the slots in a bucket might match. Actually the only reason I see why/how the tophash can help performance is cache effects.

In the chained hashtable there's no benefit to expect from an additional tophash field, as the tophash field would be part of the indexEntry and therefore end up in the same cacheline as the blob ID (well, a separate array would also be possible but that way the hashmap would always have to look at two places; grouping the topHash at the start of a batch allocation has basically the same problem. Also the chained entries are very likely each stored in a different allocation batch.). And then it's easier to simply check the first byte of the full entry ID (which is effectively identical to the tophash) and avoid spending one byte for the tophash (and there are no alignment issues). What might be worth a try is to move the `next` pointer in the `indexEntry` before the `id` field. If the id of an index entry does not match, then the hashmap only has to compare the first (few) byte(s) of the id before following the next pointer. By moving the next pointer and the start of the blob ID closer together this should reduce the chance that both values end up in different cache lines. (8 byte distance vs. 32 byte).

greatroar commented:
No significant difference. Note that comparing two ids may not be done byte-for-byte due to vectorization.

MichaelEischer commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,192 @@
+	return nil
+}
+
+var cpus = runtime.GOMAXPROCS(0)
```
I don't see much speed-up beyond two cpu cores. The speed-up is from 940ms to 650ms (for two cores) and 570ms (four cpu cores).
@greatroar Can you check which results `-cpu 1,2,4,8` as test parameters produces?

greatroar replied:
```
name                  time/op
IndexAlloc            1.16s ± 7%
IndexAlloc-2          765ms ± 2%
IndexAlloc-4          650ms ± 3%
IndexAlloc-8          620ms ± 2%
IndexAllocParallel    1.16s ± 6%
IndexAllocParallel-2  840ms ± 4%
IndexAllocParallel-4  678ms ± 3%
IndexAllocParallel-8  638ms ± 4%
```

You're right, this doesn't scale at all and it causes overhead in an already parallel setting. I'll remove it. Is a rebase ok?

greatroar replied:
Removed by git revert. The history is a mess, but I'll rebase when we're done.

MichaelEischer reviewed (commented):
The overall hashmap design seems to be settled by now. The two main open points I see right now is regarding the amount of parallelism in `grow` and the `entryAllocBatch` size.

In my opinion the `indexMap` should get a few simple tests for `add`, `len`, `foreach`, `foreachWithID` and `get`. The Index and MasterIndex test probably already cover most of the code, but I'm not sure whether these would catch all problems in the indexMap.

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,192 @@
+	key0, key1 uint64 // Key for hash randomization.
+
+	free *indexEntry // Free list.
+	wg   sync.WaitGroup
```
Multiple goroutines cannot share pointers to each other's stacks:

```
./indexmap.go:125:6: moved to heap: wg
```

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,192 @@
+	m.wg.Add(cpus)
+	for cpu := 0; cpu < cpus; cpu++ {
+		go func(cpu int) {
+			for i := cpu; i < len(old); i += cpus {
```
That's only true if they write to the same cacheline. Sharing for reading is fine, and indexMap.take only reads pointers from old, which is then left for the GC to reclaim. The writes occur inside the indexEntries, which are randomly scattered across buckets. If there's false sharing, it'll be because of entryAllocBatch and the fact that an indexEntry is slightly smaller than a cache line.

greatroar commented on internal/repository/index_test.go:
```diff
@@ -386,23 +393,23 @@ func TestIndexPacks(t *testing.T) {
+func NewRandomTestID() restic.ID {
 	id := restic.ID{}
-	rng.Read(id[:])
+	rand.Read(id[:])
```
The size and structure are already generated by the global RNG, it's only the id generation that's changed.

You're right about the lock contention, though. Profiling BenchmarkIndexAllocParallel with -count=5, -cpu=8 shows

```
  flat  flat%   sum%        cum   cum%
1850ms 10.32% 10.32%     3730ms 20.80%  runtime.scanobject
1530ms  8.53% 18.85%     3600ms 20.08%  github.com/restic/restic/internal/repository.(*indexMap).add
1340ms  7.47% 26.32%     1520ms  8.48%  runtime.findObject
1210ms  6.75% 33.07%     1210ms  6.75%  github.com/dchest/siphash.Hash
1090ms  6.08% 39.15%     1730ms  9.65%  github.com/restic/restic/internal/repository.(*indexMap).take
 780ms  4.35% 43.50%      780ms  4.35%  runtime.procyield
 730ms  4.07% 47.57%     1050ms  5.86%  math/rand.read
 700ms  3.90% 51.48%      700ms  3.90%  runtime.memclrNoHeapPointers
 690ms  3.85% 55.33%      720ms  4.02%  sync.(*Mutex).Unlock
 680ms  3.79% 59.12%     1550ms  8.64%  sync.(*Mutex).lockSlow
```

greatroar replied:
Reverted to passing the RNG in and used that for all the randomness. Also changed the way the lookupID is selected to make it robust but use only half the random numbers.

greatroar commented on internal/repository/indexmap.go:
```diff
@@ -0,0 +1,192 @@
+//
+// The buckets in this hash table contain only pointers, rather than inlined
+// key-value pairs like the standard Go map. This way, only a pointer array
+// needs to be resized when the table grows, preventing memory spikes.
```
Fixed.

greatroar commented on internal/repository/index_test.go:
```diff
@@ -330,14 +330,21 @@ func TestIndexUnserialize(t *testing.T) {
 	}
 }
 
+func BenchmarkDecodeIndexParallel(b *testing.B) {
```
Changed it (and BenchmarkDecodeIndex) to use createRandomIndex.

MichaelEischer commented on internal/repository/index_test.go:
```diff
@@ -329,7 +330,20 @@ func TestIndexUnserialize(t *testing.T) {
+	benchmarkIndexJSONOnce.Do(initBenchmarkIndexJSON)
 	b.ResetTimer()
 
 	for i := 0; i < b.N; i++ {
```
s/docExample/benchmarkIndexJSON/
for the line below

greatroar replied:
Done.

MichaelEischer reviewed (commented):
There are a few more tiny test issues that need addressing. But besides those, the only remaining issue which I want to address before merging is the `entryAllocBatch` size. I'm currently running a few more benchmark. The results currently look like most runtime variations and unexpected slowdowns between 8 or 256 disappear when I increase pass `-benchtime 10s` to the benchmark, which indicates that the different allocation sizes trigger different GC behavior/interleaving with the running code.

greatroar commented on internal/repository/indexmap_test.go:
```diff
@@ -0,0 +1,133 @@
+	rtest.Equals(t, len(ids), n)
+
+	n = 0
+	m.foreach(func(*indexEntry) bool { return false })
```
Good catch, fixed.

greatroar commented on internal/repository/indexmap_test.go:
```diff
@@ -0,0 +1,133 @@
+
+	n := 0
+	m.foreach(func(e *indexEntry) bool {
+		_, ok := ids[e.id]
```
Fixed in a somewhat simpler way.

greatroar commented:
I just reran the latest benchmarks (fast createRandomIndex, proper size for DecodeIndex*) with entryAllocBatch = 8 as old and 256 as new. These are the results:

```
name                                            old time/op    new time/op    delta
IndexAlloc-8                                       688ms ± 8%     656ms ± 5%   -4.68%  (p=0.006 n=20+20)
IndexAllocParallel-8                               306ms ±16%     233ms ±14%  -23.73%  (p=0.000 n=20+20)
IndexMapHash-8                                    3.61µs ± 0%    3.61µs ± 1%     ~     (p=0.484 n=9+10)
PackerManager-8                                    181ms ± 1%     179ms ± 0%   -0.77%  (p=0.000 n=10+9)
DecodeIndex-8                                      4.34s ± 0%     4.34s ± 1%     ~     (p=0.604 n=9+10)
DecodeIndexParallel-8                              4.34s ± 0%     4.34s ± 1%     ~     (p=0.842 n=10+9)
IndexHasUnknown-8                                 49.4ns ±12%    53.0ns ±16%     ~     (p=0.196 n=10+10)
IndexHasKnown-8                                   54.3ns ±17%    51.7ns ± 9%     ~     (p=0.210 n=10+10)
MasterIndexLookupSingleIndex-8                     144ns ± 5%     138ns ± 3%   -4.38%  (p=0.007 n=10+9)
MasterIndexLookupMultipleIndex-8                   419ns ± 6%     424ns ± 4%     ~     (p=0.424 n=10+10)
MasterIndexLookupSingleIndexUnknown-8             73.5ns ± 9%    70.7ns ± 8%     ~     (p=0.305 n=10+9)
MasterIndexLookupMultipleIndexUnknown-8            351ns ±10%     353ns ± 3%     ~     (p=0.328 n=9+9)
MasterIndexLookupParallel/known,indices=5-8        859ns ± 2%     879ns ± 5%   +2.30%  (p=0.011 n=10+10)
MasterIndexLookupParallel/unknown,indices=5-8      278ns ± 2%     295ns ± 1%   +6.31%  (p=0.000 n=10+8)
MasterIndexLookupParallel/known,indices=10-8      1.17µs ± 1%    1.14µs ± 1%   -2.57%  (p=0.000 n=10+10)
MasterIndexLookupParallel/unknown,indices=10-8     261ns ± 1%     253ns ± 1%   -3.07%  (p=0.000 n=10+10)
MasterIndexLookupParallel/known,indices=20-8      1.29µs ± 0%    1.21µs ± 2%   -5.71%  (p=0.000 n=7+10)
MasterIndexLookupParallel/unknown,indices=20-8     564ns ± 1%     533ns ± 1%   -5.63%  (p=0.000 n=10+10)
SaveAndEncrypt-8                                  98.8ns ± 1%    98.5ns ± 0%     ~     (p=0.348 n=9+9)
LoadTree-8                                         577µs ± 1%     575µs ± 0%   -0.50%  (p=0.011 n=9+9)
LoadBlob-8                                        5.70ms ± 2%    5.72ms ± 2%     ~     (p=0.280 n=10+10)
LoadAndDecrypt-8                                  5.93ms ± 0%    5.93ms ± 0%     ~     (p=0.243 n=9+10)
LoadIndex-8                                       24.7ms ± 0%    24.6ms ± 1%   -0.50%  (p=0.003 n=10+10)

name                                            old alloc/op   new alloc/op   delta
IndexAlloc-8                                       401MB ± 0%     401MB ± 0%   +0.00%  (p=0.000 n=18+19)
IndexAllocParallel-8                               400MB ± 0%     400MB ± 0%   +0.00%  (p=0.000 n=20+20)
IndexMapHash-8                                     0.00B          0.00B          ~     (all equal)
PackerManager-8                                   91.6kB ± 0%    91.6kB ± 0%     ~     (p=0.291 n=9+9)
SaveAndEncrypt-8                                   1.00B ± 0%     1.00B ± 0%     ~     (all equal)

name                                            old allocs/op  new allocs/op  delta
IndexAlloc-8                                       1.12M ± 0%     0.91M ± 0%  -18.79%  (p=0.000 n=19+18)
IndexAllocParallel-8                               1.12M ± 0%     0.91M ± 0%  -18.79%  (p=0.000 n=20+20)
IndexMapHash-8                                      0.00           0.00          ~     (all equal)
PackerManager-8                                    1.41k ± 0%     1.41k ± 0%     ~     (p=0.173 n=10+10)
SaveAndEncrypt-8                                    0.00           0.00          ~     (all equal)

name                                            old speed      new speed      delta
IndexMapHash-8                                  1.13GB/s ± 0%  1.13GB/s ± 1%     ~     (p=0.497 n=9+10)
PackerManager-8                                  291MB/s ± 1%   294MB/s ± 0%   +0.77%  (p=0.000 n=10+9)
SaveAndEncrypt-8                                42.5TB/s ± 1%  42.6TB/s ± 0%     ~     (p=0.297 n=9+9)
LoadBlob-8                                       176MB/s ± 2%   175MB/s ± 2%     ~     (p=0.280 n=10+10)
LoadAndDecrypt-8                                 169MB/s ± 0%   169MB/s ± 0%     ~     (p=0.242 n=9+10)
```

In summary, constructing indexes gets faster, while lookups get a bit slower. Still nothing that merged indexes couldn't solve, I think, given that the loss is mainly in the lookups of unknown blobs. The overallocation is statistically significant but still tiny.

MichaelEischer commented:
I've also run a few of the lookup benchmarks with `-benchtime 10s` and for those there's nearly no difference between a batchsize of 8 or 256, with slightly better performance for a batchsize of 256.

```
name                                            old time/op  new time/op  delta
MasterIndexLookupSingleIndex-8                   134ns ± 6%   131ns ± 6%    ~     (p=0.173 n=20+20)
MasterIndexLookupMultipleIndex-8                 376ns ± 7%   364ns ± 7%  -3.26%  (p=0.011 n=19+19)
MasterIndexLookupSingleIndexUnknown-8           60.2ns ±11%  60.5ns ±11%    ~     (p=0.700 n=20+20)
MasterIndexLookupMultipleIndexUnknown-8          285ns ± 7%   287ns ±13%    ~     (p=0.808 n=19+20)
MasterIndexLookupParallel/known,indices=5-8      342ns ± 5%   340ns ± 5%    ~     (p=0.440 n=20+20)
MasterIndexLookupParallel/unknown,indices=5-8    232ns ± 8%   232ns ± 6%    ~     (p=0.995 n=20+20)
MasterIndexLookupParallel/known,indices=10-8     684ns ± 7%   673ns ± 4%    ~     (p=0.122 n=20+20)
MasterIndexLookupParallel/unknown,indices=10-8   318ns ± 4%   316ns ± 9%    ~     (p=0.399 n=20+19)
MasterIndexLookupParallel/known,indices=20-8     720ns ± 8%   707ns ± 9%    ~     (p=0.168 n=20+19)
MasterIndexLookupParallel/unknown,indices=20-8   487ns ± 4%   490ns ± 4%    ~     (p=0.285 n=20+20)
```

So I'd say we use `256` as batch size, which also has the benefit of a slightly lower memory usage.
@greatroar Do you want to rebase / squash / clean up some of the commits before merging? Or did I miss any points which still need discussion?

greatroar commented:
Switched to 256, rebased everything and re-ran all the benchmarks (see commit messages; this took several hours). If the tests pass, I'm done.

MichaelEischer reviewed (approved):
LGTM. Thanks a lot @greatroar and @aawsome for all the discussions and work on this!

greatroar commented:
Thank you both for the thorough review!
